lib/report.js

```diff
@@ -158,7 +160,123 @@ class ConciseReporter extends Reporter {
   }
 }
 
+class TapReporter extends Reporter {
+  constructor(options) {
+    super(options);
+    this.successful = 0;
+    this.failed = 0;
+    this.skipped = 0;
+    this.failedTests = [];
+    this.testNumeration = 1;
+    this.message = '';
+    this.got = [];
+    this.expected = [];
+  }
+
+  error(test, error) {
+    // remove one success because 'error' may only be received after end
+    // though the test may have already failed so we need a way to check
+    // if we should do this, i.e. store map of successful tests
+    // this.successes--;
+    // test.todo ? this.skipped++ : this.failures++;
+    // TODO(lundibundi): add test.metadata.filename here too once available
+    this.listFailure(
+      test,
+      { message: `Failure after end: ${error ? error.message : 'no message'}` },
+      ` Test failed: ${test.caption}`
+    );
+  }
+
+  parseTestResults(test, offset = 0) {
+    test.results.forEach((res, index) => {
+      const testId = index + 1;
+      if (res.type === 'subtest') {
+        const t = res.test;
+        console.log(`${pad(offset + 2)}# Subtest: ${t.caption}`);
+        this.parseTestResults(t, offset + 2);
+        const prefix = t.success ? 'ok' : 'not ok';
+        console.log(`${pad(offset + 2)}${prefix} ${testId} - ${t.caption}`);
+      } else {
```

Comment:
Since you do not assign the result of execution to some variable, it will be better to use `if-else`.
```suggestion
      } else if (res.success) {
        this.checkSuccess(res, testId, offset + 2)
      } else {
        this.checkFailure(res, testId, offset + 2)
```
Alternatively, you can return `message` from `checkSuccess` and `listFailure` and print them here.
```suggestion
      } else {
        const message = res.success
           ? this.checkSuccess(res, testId, offset + 2)
           : this.listFailure(res, testId, offset + 2);
         console.log(message);
      }
```